server/user.go

```diff
@@ -0,0 +1,201 @@
+package main
+
+import (
+	"bytes"
+	"context"
+	"crypto/aes"
+	"crypto/cipher"
+	"crypto/rand"
+	"encoding/base64"
+	"encoding/json"
+	"io"
+
+	"github.com/pkg/errors"
+	msgraph "github.com/yaegashi/msgraph.go/beta"
+	"golang.org/x/oauth2"
+)
+
+// UserInfo defines the information we store from each user
+type UserInfo struct {
+	Email      string
+	OAuthToken *oauth2.Token
+	// Mattermost userID
+	UserID string
+	// Remote userID
+	RemoteID string
+	// Remote UPN
+	UPN string
+}
+
+const (
+	tokenKey           = "token_"
+	tokenKeyByRemoteID = "tbyrid_"
+)
+
+func (c *Client) GetMe() (*msgraph.User, error) {
+	ctx := context.Background()
+	graphUser, err := c.builder.Me().Request().Get(ctx)
+	if err != nil {
+		c.api.LogError(errors.Wrap(err, "cannot get user").Error())
+		return nil, err
+	}
+
+	if graphUser == nil {
+		err = errors.New("empty user")
+		c.api.LogError(errors.Wrap(err, "cannot get user").Error())
+		return nil, err
+	}
+
+	return graphUser, nil
+}
+
+func (p *Plugin) StoreUserInfo(info *UserInfo) error {
+	encryptionKey := p.getConfiguration().EncryptionKey
+	if encryptionKey != "" {
+		token := info.OAuthToken.AccessToken
+		encryptedToken, err := encrypt([]byte(encryptionKey), token)
+		if err != nil {
+			return errors.Wrap(err, "error occurred while encrypting access token")
+		}
+		info.OAuthToken.AccessToken = encryptedToken
+		p.API.LogDebug("encrypted user token", "encrypted", lastN(encryptedToken, 4), "token", lastN(token, 4))
+	}
+
+	jsonInfo, err := json.Marshal(info)
+	if err != nil {
+		return err
+	}
+
+	if err := p.API.KVSet(tokenKey+info.UserID, jsonInfo); err != nil {
+		return err
+	}
+
+	if err := p.API.KVSet(tokenKeyByRemoteID+info.RemoteID, jsonInfo); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (p *Plugin) GetUserInfo(userID string) (*UserInfo, error) {
+	var userInfo UserInfo
+
+	infoBytes, appErr := p.API.KVGet(tokenKey + userID)
+	if appErr != nil || infoBytes == nil {
+		return nil, errors.New("must connect user account to Microsoft first")
+	}
+
+	err := json.Unmarshal(infoBytes, &userInfo)
+	if err != nil {
+		return nil, errors.New("unable to parse token")
+	}
+
+	encryptionKey := p.getConfiguration().EncryptionKey
+	if encryptionKey != "" {
+		encryptedToken := userInfo.OAuthToken.AccessToken
+		token, err := decrypt([]byte(encryptionKey), encryptedToken)
+		if err != nil {
+			p.API.LogWarn("Failed to decrypt access token", "error", err.Error())
+			return nil, errors.Wrap(err, "failed to decrypt previously stored user access token")
+		}
+
+		userInfo.OAuthToken.AccessToken = token
+		p.API.LogDebug("decrypted user token", "encrypted", lastN(encryptedToken, 4), "token", lastN(token, 4))
+	}
```

Comment:
maybe add a `p.API.LogDebug/Info` statement to capture that we don't have an encryption key? (might be helpful for debugging later on)